worlds/tww/randomizers/RequiredBosses.py

```diff
@@ -0,0 +1,121 @@
+            )
+
+        # Ensure that after removing excluded dungeons, we still have enough to satisfy user options.
+        num_remaining = num_required_bosses - len(required_dungeons)
+        remaining_dungeon_options = dungeon_names - required_dungeons - options.excluded_dungeons.value
+        if len(remaining_dungeon_options) < num_remaining:
+            raise OptionError(
+                "Could not select required bosses to satisfy options set by the user. "
+                "After removing the excluded dungeons, there are not enough to meet the desired number of required"
+                "bosses."
+            )
+
+        # Finish selecting required bosses.
+        required_dungeons.update(self.world.random.sample(list(remaining_dungeon_options), num_remaining))
+
+        # Exclude locations that are not in the dungeon of a required boss.
+        banned_dungeons = dungeon_names - required_dungeons
+        for location_name, location_data in LOCATION_TABLE.items():
+            dungeon_name, _ = split_location_name_by_zone(location_name)
+            if dungeon_name in banned_dungeons and TWWFlag.DUNGEON in location_data.flags:
+                self.banned_locations.add(location_name)
+            elif location_name == "Mailbox - Letter from Orca" and "Forbidden Woods" in banned_dungeons:
+                self.banned_locations.add(location_name)
+            elif location_name == "Mailbox - Letter from Baito" and "Earth Temple" in banned_dungeons:
+                self.banned_locations.add(location_name)
+            elif location_name == "Mailbox - Letter from Aryll" and "Forsaken Fortress" in banned_dungeons:
+                self.banned_locations.add(location_name)
+            elif location_name == "Mailbox - Letter from Tingle" and "Forsaken Fortress" in banned_dungeons:
+                self.banned_locations.add(location_name)
+        for location_name in self.banned_locations:
+            self.world.nonprogress_locations.add(location_name)
+
+        # Record the item location names for required bosses.
+        self.required_boss_item_locations = []
+        self.required_bosses = []
+        self.banned_bosses = []
+        possible_boss_item_locations = [loc for loc, data in LOCATION_TABLE.items() if TWWFlag.BOSS in data.flags]
+        for location_name in possible_boss_item_locations:
+            dungeon_name, specific_location_name = split_location_name_by_zone(location_name)
+            assert specific_location_name.endswith(" Heart Container")
+            boss_name = specific_location_name.removesuffix(" Heart Container")
+
+            if dungeon_name in required_dungeons:
+                self.required_boss_item_locations.append(location_name)
+                self.required_bosses.append(boss_name)
+            else:
+                self.banned_bosses.append(boss_name)
+        self.required_dungeons = list(required_dungeons)
+        self.banned_dungeons = list(banned_dungeons)
```

Comment:
Both of these lists only appear to be used like sets (`if dungeon_name in self.world.boss_reqs.banned_dungeons:` etc.), so should be sets instead of lists.

Because `required_dungeons` and `banned_dungeons` are both `set[str]`, these lists created from them have nondeterministic order, so are dangerous to keep around as is.

The type hints in `__init__` will also need updating.

```suggestion
        self.required_dungeons = required_dungeons
        self.banned_dungeons = banned_dungeons
```